clientlib/clientlib.go

```diff
@@ -85,38 +122,66 @@ func Schedule(msg *kafka.Message, scheduleID string, epoch int64, schedulerTopic
 	return &m, nil
 }
 
+// ScheduleRetry retries a Kafka message by using the specified retry configuration.
+func ScheduleRetry(msg *kafka.Message, scheduleID, schedulerTopic string, config *retry.Config) (*kafka.Message, error) {
+	if config == nil {
+		return nil, ErrNoRetryConfigProvided
+	}
+
+	attempt := setHeaderRetryAttempt(msg)
+
+	if attempt > config.MaxAttempt {
+		return nil, ErrMaxAttemptReached
+	} else if attempt == 1 {
+		currentTimestamp := []byte(strconv.FormatInt(Now().Unix(), 10))
+		updateOrAddHeader(msg, HeaderRetryOriginalTimestamp, currentTimestamp)
+	}
+
+	return Schedule(msg, scheduleID, getNextEpoch(attempt, config.Strategy), schedulerTopic)
+}
+
 // DeleteSchedule creates a message for deleting a schedule. This message goal is to be produce by a kafka producer.
 func DeleteSchedule(scheduleID, schedulerTopic string) (*kafka.Message, error) {
 	// Invalid parameters
 	if schedulerTopic == "" || scheduleID == "" {
 		return nil, fmt.Errorf("invalid parameter")
 	}
 
-	bytes := func(i string) []byte {
-		return []byte(i)
-	}
-
 	m := kafka.Message{
 		TopicPartition: kafka.TopicPartition{Topic: &schedulerTopic, Partition: kafka.PartitionAny},
-		Key:            bytes(scheduleID),
+		Key:            []byte(scheduleID),
 		// tombstone message has nil Value
 		Value: nil,
 	}
 
 	return &m, nil
 }
 
-// IsSchedulerMessage tells if a message is coming from the scheduler
+// IsSchedulerMessage tells if a message is coming from the scheduler.
 func IsSchedulerMessage(msg *kafka.Message) bool {
-	if msg == nil {
-		return false
-	}
+	return getHeader(msg, SchedulerKey).Key == SchedulerKey
+}
 
-	for _, h := range msg.Headers {
-		if h.Key == SchedulerKey {
-			return true
-		}
+// IsRetriedMessage tells if a message has already been retried by the scheduler.
+func IsRetriedMessage(msg *kafka.Message) bool {
+	return getHeader(msg, HeaderRetryAttempt).Key == HeaderRetryAttempt
+}
+
+func setHeaderRetryAttempt(msg *kafka.Message) int {
+	attempt := 1
+	attemptHeader := getHeader(msg, HeaderRetryAttempt)
+
+	if attemptHeader.Key == HeaderRetryAttempt {
```

Comment:
you dont have to check the key, because you got the header from line 172.
Or you can get directly the value with getHeaderValue

Reply:
Thank you, I am now using the `getHeaderValue()` function